src/main/java/org/opensearch/security/dlic/rest/api/AuthFailureListenersApiAction.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.dlic.rest.api;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.security.DefaultObjectMapper;
+import org.opensearch.security.dlic.rest.validation.EndpointValidator;
+import org.opensearch.security.dlic.rest.validation.RequestContentValidator;
+import org.opensearch.security.dlic.rest.validation.RequestContentValidator.DataType;
+import org.opensearch.security.securityconf.impl.CType;
+import org.opensearch.security.securityconf.impl.v7.ConfigV7;
+import org.opensearch.security.support.SecurityJsonNode;
+import org.opensearch.threadpool.ThreadPool;
+
+import static org.opensearch.rest.RestRequest.Method.DELETE;
+import static org.opensearch.rest.RestRequest.Method.GET;
+import static org.opensearch.rest.RestRequest.Method.PUT;
+import static org.opensearch.security.dlic.rest.api.Responses.badRequest;
+import static org.opensearch.security.dlic.rest.api.Responses.ok;
+import static org.opensearch.security.dlic.rest.api.Responses.response;
+import static org.opensearch.security.dlic.rest.support.Utils.addRoutesPrefix;
+
+public class AuthFailureListenersApiAction extends AbstractApiAction {
+
+    public static final String NAME_JSON_PROPERTY = "name";
+
+    public static final String TYPE_JSON_PROPERTY = "type";
+    public static final String AUTHENTICATION_BACKEND_JSON_PROPERTY = "authentication_backend";
+    public static final String ALLOWED_TRIES_JSON_PROPERTY = "allowed_tries";
+    public static final String TIME_WINDOW_SECONDS_JSON_PROPERTY = "time_window_seconds";
+    public static final String BLOCK_EXPIRY_JSON_PROPERTY = "block_expiry_seconds";
+    public static final String MAX_BLOCKED_CLIENTS_JSON_PROPERTY = "max_blocked_clients";
+    public static final String MAX_TRACKED_CLIENTS_JSON_PROPERTY = "max_tracked_clients";
+
+    private static final List<Route> ROUTES = addRoutesPrefix(
+        ImmutableList.of(
+            new Route(GET, "/authfailurelisteners"),
+            new Route(DELETE, "/authfailurelisteners/{name}"),
+            new Route(PUT, "/authfailurelisteners/{name}")
```

Comment:
nit: can we make this more user-friendly? Also, in opensearch docs, the feature is called API rate limiting, should the names match the documentation? (while I see that the config name is ``auth_failure_listeners`` ) - https://opensearch.org/docs/latest/security/configuration/api-rate-limiting/

Reply:
Is the use-case to limit API calls based on this config? if so, it should be renamed to `/api-rate-limiter`

Reply:
Are more than one auth_failure_listener allowed? My understanding is there can be at most 1 ip-based and 1 username-based.

If its limited to one, should we not accept a url param and have separate routes for setting up ip-based and username-based? If `{name}` is an accepted value then would the acceptable values be `ip` and `username`?

If this `{name}` is used for the key in yaml/security index for the auth_failure_listener (i.e. `ip_rate_limiting` or `internal_authentication_backend_limiting` in the demo configuration), then I don't think its necessary to provide this kind of customization. If they can only configure a single `auth_failure_listener` per type then we should keep the `ip_rate_limiting` and `internal_authentication_backend_limiting` convention

Reply:
I took a look through the code and it seems like there can be more than one per type: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/DynamicConfigModelV7.java#L454. I don't know if this makes sense though, since I don't really see a scenario that you would want multiple of each type. @cwperks should we file a bug around this?

Reply:
As to the other points in this thread, I was thinking rate limiting might be a misleading term to brand these APIs, since they are only limiting on auth failures, not on all API requests. I opened an issue to see if there is interest in the bigger feature of rate limiting all API requests (regardless of auth failure) [here](https://github.com/opensearch-project/security/issues/4652). I can change it if the majority thinks otherwise.

Reply:
This matches the same pattern for [InternalUsers API](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java#L74-L81) and [RolesMapping API](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/dlic/rest/api/RolesMappingApiAction.java#L39-L48). I think we consider different names, but I also think the name in this PR is fine for consistency.